File: app/src/main/java/au/com/codeka/weather/WeatherDetailsFragment.java
package au.com.codeka.weather;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.florent37.materialviewpager.MaterialViewPagerHelper;
import com.github.ksoichiro.android.observablescrollview.ObservableScrollView;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Locale;

import au.com.codeka.weather.model.CurrentCondition;
import au.com.codeka.weather.model.Forecast;
import au.com.codeka.weather.model.HourlyForecast;
import au.com.codeka.weather.model.WeatherInfo;

/**
 * Fragment which actually displays the details about the weather.
 */
public class WeatherDetailsFragment extends Fragment {
  private View rootView;
  private ObservableScrollView scrollView;

  @Override
  public View onCreateView(
      LayoutInflater inflater,
      ViewGroup container,
      Bundle savedInstanceState) {
    rootView = inflater.inflate(
        R.layout.weather_details_fragment, container, false);
    scrollView = (ObservableScrollView) rootView.findViewById(R.id.scroll_view);

    return rootView;
  }

  @Override
  public void onStart() {
    super.onStart();
    WeatherManager.i.addUpdateRunnable(weatherUpdatedRunnable);
    refresh();
  }

  public void onStop() {
    super.onStop();
    WeatherManager.i.removeUpdateRunnable(weatherUpdatedRunnable);
  }

  private void refresh() {
    WeatherInfo weatherInfo = WeatherManager.i.getCurrentWeather(getActivity());
    if (weatherInfo == null) {
      return;
    }

    CurrentCondition currentCondition = weatherInfo.getCurrentCondition();
    if (currentCondition == null) {
      return;
    }

    // a bit hacky...
    int hour = new GregorianCalendar().get(Calendar.HOUR_OF_DAY);
    boolean isNight = hour < 6 || hour > 20;

    LayoutInflater inflater = LayoutInflater.from(getActivity());

    ((ImageView) rootView.findViewById(R.id.current_icon)).setImageResource(
        currentCondition.getIcon().getLargeIconId(isNight));
    if (currentCondition.getTemperature() == null) {
      ((TextView) rootView.findViewById(R.id.current_temp)).setText("??°C");
    } else {
      ((TextView) rootView.findViewById(R.id.current_temp)).setText(
          String.format("%d°C", Math.round(currentCondition.getTemperature())));
    }
    ((TextView) rootView.findViewById(R.id.current_description)).setText(
        currentCondition.getDescription());
    ((TextView) rootView.findViewById(R.id.observation_location)).setText(
        String.format("at %s", currentCondition.getObservationLocation()));
    Date now = new Date();
    long millis = now.getTime() - currentCondition.getObservationTime().getTime();
    ((TextView) rootView.findViewById(R.id.observation_time)).setText(
        String.format("%d minutes ago", millis / 60000L));

    String feelsLike = currentCondition.getFeelsLike() == null
        ? "??"
        : Long.toString(Math.round(currentCondition.getFeelsLike()));
    String relativeHumidity = currentCondition.getRelativeHumidity() == null
        ? "??"
        : Long.toString(Math.round(currentCondition.getRelativeHumidity()));
    ((TextView) rootView.findViewById(R.id.extra_info_1)).setText(
        String.format("Feels like %s°C, %s%% relative humidity", feelsLike, relativeHumidity));
    ((TextView) rootView.findViewById(R.id.extra_info_2)).setText(
        String.format("%dmm last hour, %dmm today",
            Math.round(currentCondition.getPrecipitationLastHour()),
            Math.round(currentCondition.getPrecipitationToday())));

    LinearLayout hourlyParent = (LinearLayout) rootView.findViewById(R.id.hourly_container);
    hourlyParent.removeAllViews();
    for (HourlyForecast hourlyForecast : weatherInfo.getHourlyForecasts()) {
      View hourlyForecastView =
          inflater.inflate(R.layout.weather_details_hourly_row, hourlyParent, false);

      isNight = hourlyForecast.getHour() < 6 || hourlyForecast.getHour() > 20;
      String hourValue = Integer.toString(hourlyForecast.getHour()) + "am";
      if (hourlyForecast.getHour() == 0) {
        hourValue = "12am";
      } else if (hourlyForecast.getHour() == 12) {
        hourValue = "12pm";
      } else if (hourlyForecast.getHour() > 12) {
        hourValue = Integer.toString(hourlyForecast.getHour() - 12) + "pm";
      }
      ((TextView) hourlyForecastView.findViewById(R.id.forecast_hour)).setText(hourValue);
      ((ImageView) hourlyForecastView.findViewById(R.id.forecast_icon)).setImageResource(
          hourlyForecast.getIcon().getSmallIconId(isNight));
      ((TextView) hourlyForecastView.findViewById(R.id.forecast_temp)).setText(
          String.format("%d°C", Math.round(hourlyForecast.getTemperature())));
      if (hourlyForecast.getQpfMillimeters() < 0.5) {
        hourlyForecastView.findViewById(R.id.forecast_precipitation).setVisibility(View.GONE);
      } else {
        hourlyForecastView.findViewById(R.id.forecast_precipitation).setVisibility(View.VISIBLE);
        ((TextView) hourlyForecastView.findViewById(R.id.forecast_precipitation)).setText(
            String.format("%dmm", Math.round(hourlyForecast.getQpfMillimeters())));
      }

      hourlyParent.addView(hourlyForecastView);
    }

    CardLinearLayout forecastParent =
        (CardLinearLayout) rootView.findViewById(R.id.weather_cards);
    forecastParent.removeAllViews();
    for (Forecast forecast : weatherInfo.getForecasts()) {
      View forecastView = inflater.inflate(
          R.layout.weather_details_forecast_row, forecastParent, false);

      String day = "Today";
      if (forecast.getOffset() == 1) {
        day = "Tomorrow";
      } else if (forecast.getOffset() > 1) {
        Calendar cal = new GregorianCalendar();
        cal.add(Calendar.DAY_OF_YEAR, forecast.getOffset());
        day = cal.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.LONG, Locale.US);
      }
      ((TextView) forecastView.findViewById(R.id.forecast_day)).setText(day);
      ((TextView) forecastView.findViewById(R.id.forecast_description)).setText(
          forecast.getDescription());
      ((TextView) forecastView.findViewById(R.id.forecast_temp_high)).setText(
          String.format("%d°C", Math.round(forecast.getHighTemperature())));
      ((TextView) forecastView.findViewById(R.id.forecast_temp_low)).setText(
          String.format("%d°C", Math.round(forecast.getLowTemperature())));
      ((ImageView) forecastView.findViewById(R.id.forecast_icon)).setImageResource(
          forecast.getIcon().getLargeIconId(false));

      forecastParent.addView(forecastView);
    }
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    MaterialViewPagerHelper.registerScrollView(getActivity(), scrollView, null);
  }

  // Called when the weather updates.
  private final Runnable weatherUpdatedRunnable = new Runnable() {
    @Override
    public void run() {
      getActivity().runOnUiThread(new Runnable() {
        @Override
        public void run() {
          refresh();
        }
      });
    }
  };
}

File: app/src/main/java/au/com/codeka/weather/WeatherDetailsFragment.kt
package au.com.codeka.weather

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.HorizontalScrollView
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import au.com.codeka.weather.model.WeatherInfo
import com.github.florent37.materialviewpager.MaterialViewPagerHelper
import java.util.*
import kotlin.math.roundToInt

/**
 * Fragment which actually displays the details about the weather.
 */
class WeatherDetailsFragment : Fragment() {
  private var rootView: View? = null
  private var scrollView: NestedScrollView? = null
  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    rootView = inflater.inflate(
        R.layout.weather_details_fragment, container, false)
    scrollView = rootView!!.findViewById(R.id.scroll_view)
    return rootView
  }

  override fun onStart() {
    super.onStart()
    WeatherManager.Companion.i.addUpdateRunnable(weatherUpdatedRunnable)
    refresh()
  }

  override fun onStop() {
    super.onStop()
    WeatherManager.Companion.i.removeUpdateRunnable(weatherUpdatedRunnable)
  }

  private fun refresh() {
    val weatherInfo: WeatherInfo = WeatherManager.Companion.i.getCurrentWeather(activity) ?: return
    val currentCondition = weatherInfo.currentCondition ?: return

    // a bit hacky...
    val hour = GregorianCalendar()[Calendar.HOUR_OF_DAY]
    var isNight = hour < 6 || hour > 20
    val inflater = LayoutInflater.from(activity)
    (rootView!!.findViewById<View>(R.id.current_icon) as ImageView).setImageResource(
        currentCondition.icon!!.getLargeIconId(isNight))
    if (currentCondition.temperature == null) {
      (rootView!!.findViewById<View>(R.id.current_temp) as TextView).text = "??°C"
    } else {
      (rootView!!.findViewById<View>(R.id.current_temp) as TextView).text = String.format("%d°C", (currentCondition.temperature
          ?: 0.0).roundToInt())
    }
    (rootView!!.findViewById<View>(R.id.current_description) as TextView).text = currentCondition.description
    (rootView!!.findViewById<View>(R.id.observation_location) as TextView).text = String.format("at %s", currentCondition.observationLocation)
    val now = Date()
    val millis = now.time - currentCondition.observationTime!!.time
    (rootView!!.findViewById<View>(R.id.observation_time) as TextView).text = String.format("%d minutes ago", millis / 60000L)
    val feelsLike = if (currentCondition.feelsLike == null) "??" else (currentCondition.feelsLike ?: 0.0).roundToInt().toString()
    val relativeHumidity = if (currentCondition.relativeHumidity == null) "??" else (currentCondition.relativeHumidity ?: 0.0).roundToInt().toString()
    (rootView!!.findViewById<View>(R.id.extra_info_1) as TextView).text = String.format("Feels like %s°C, %s%% relative humidity", feelsLike, relativeHumidity)
    (rootView!!.findViewById<View>(R.id.extra_info_2) as TextView).text = String.format("%dmm last hour, %dmm today",
        currentCondition.precipitationLastHour.roundToInt(),
        currentCondition.precipitationToday.roundToInt())
    val hourlyParent = rootView!!.findViewById<View>(R.id.hourly_container) as LinearLayout
    hourlyParent.removeAllViews()
    for (hourlyForecast in weatherInfo.hourlyForecasts) {
      val hourlyForecastView = inflater.inflate(R.layout.weather_details_hourly_row, hourlyParent, false)
      isNight = hourlyForecast.hour < 6 || hourlyForecast.hour > 20
      var hourValue = hourlyForecast.hour.toString() + "am"
      if (hourlyForecast.hour == 0) {
        hourValue = "12am"
      } else if (hourlyForecast.hour == 12) {
        hourValue = "12pm"
      } else if (hourlyForecast.hour > 12) {
        hourValue = Integer.toString(hourlyForecast.hour - 12) + "pm"
      }
      (hourlyForecastView.findViewById<View>(R.id.forecast_hour) as TextView).text = hourValue
      (hourlyForecastView.findViewById<View>(R.id.forecast_icon) as ImageView).setImageResource(hourlyForecast.icon?.getSmallIconId(isNight) ?: 0)
      (hourlyForecastView.findViewById<View>(R.id.forecast_temp) as TextView).text = String.format("%d°C", hourlyForecast.temperature.roundToInt())
      if (hourlyForecast.qpfMillimeters < 0.5) {
        hourlyForecastView.findViewById<View>(R.id.forecast_precipitation).visibility = View.GONE
      } else {
        hourlyForecastView.findViewById<View>(R.id.forecast_precipitation).visibility = View.VISIBLE
        (hourlyForecastView.findViewById<View>(R.id.forecast_precipitation) as TextView).text = String.format("%dmm", hourlyForecast.qpfMillimeters.roundToInt())
      }
      hourlyParent.addView(hourlyForecastView)
    }
    val forecastParent: CardLinearLayout = rootView!!.findViewById(R.id.weather_cards)
    // Remove any views after the HorizontalScrollView, since they'll be old entries
    for (i in 0 until forecastParent.childCount) {
      if (forecastParent.getChildAt(i) is HorizontalScrollView) {
        val j = i + 1
        while (j < forecastParent.childCount) {
          forecastParent.removeViewAt(j)
        }
        break
      }
    }
    for (forecast in weatherInfo.forecasts) {
      val forecastView = inflater.inflate(
          R.layout.weather_details_forecast_row, forecastParent, false)
      var day = "Today"
      if (forecast.offset == 1) {
        day = "Tomorrow"
      } else if (forecast.offset > 1) {
        val cal: Calendar = GregorianCalendar()
        cal.add(Calendar.DAY_OF_YEAR, forecast.offset)
        day = cal.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.LONG, Locale.US)
      }
      (forecastView.findViewById<View>(R.id.forecast_day) as TextView).text = day
      (forecastView.findViewById<View>(R.id.forecast_description) as TextView).text = forecast.description
      (forecastView.findViewById<View>(R.id.forecast_temp_high) as TextView).text = String.format("%d°C", forecast.highTemperature.roundToInt())
      (forecastView.findViewById<View>(R.id.forecast_temp_low) as TextView).text = String.format("%d°C", forecast.lowTemperature.roundToInt())
      (forecastView.findViewById<View>(R.id.forecast_icon) as ImageView).setImageResource(
          forecast.icon!!.getLargeIconId(false))
      forecastParent.addView(forecastView)
    }
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    MaterialViewPagerHelper.registerScrollView(context, scrollView)
  }

  // Called when the weather updates.
  private val weatherUpdatedRunnable = Runnable { activity!!.runOnUiThread { refresh() } }
}